Federated Model (seq-to-seq)/train_model.py

```diff
@@ -0,0 +1,358 @@
+    # that your model expects.
+    keras_model = create_keras_model(arg, input_vocab_size, slot_vocab_size,
+                                     intent_vocab_size)
+    return tff.learning.from_keras_model(
+        keras_model,
+        input_spec=input_spec,
+        loss=[
+            masked_slot_loss,
+            tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+        ])
+
+
+def main():
+    """Runs the entire pipelone from loading the data and defining the model
+    to training the model and evaluating the model.
+    """
+
+    arg = parse_arguments()
+
+    in_vocab, slot_vocab, intent_vocab = load_vocab(arg)
+
+    # Loading data
+    train_in_data, train_slot_data, train_intent_data = load_dataset(
+        arg, arg.train_data_path, in_vocab, slot_vocab, intent_vocab)
+
+    # valid_in_data, valid_slot_data, valid_intent_data = load_dataset(
+    #     arg, arg.valid_data_path, in_vocab, slot_vocab, intent_vocab)
+
+    # test_in_data, test_slot_data, test_intent_data = load_dataset(
+    #     arg, arg.test_data_path, in_vocab, slot_vocab, intent_vocab)
+
+    targets = train_slot_data[:, 1:]
+    padding_masks, look_ahead_masks, intent_masks = create_masks(
+        train_in_data, targets)
+
+    num_clients = 10
+    ftrain_data = collections.OrderedDict()
+    instance_per_client = len(train_in_data) // num_clients
+
+    for i in range(num_clients):
+        client_data = collections.OrderedDict(
+            x=(train_in_data[i * instance_per_client:(i + 1) *
+                             instance_per_client],
+               train_slot_data[i * instance_per_client:(i + 1) *
+                               instance_per_client, 1:],
+               padding_masks[i * instance_per_client:(i + 1) *
+                             instance_per_client],
+               look_ahead_masks[i * instance_per_client:(i + 1) *
+                                instance_per_client],
+               intent_masks[i * instance_per_client:(i + 1) *
+                            instance_per_client]),
+            y=(targets[i * instance_per_client:(i + 1) * instance_per_client],
+               train_intent_data[i * instance_per_client:(i + 1) *
+                                 instance_per_client]))
+        ftrain_data[str(i)] = client_data
+
+    ftrain_data = tff.simulation.FromTensorSlicesClientData(ftrain_data)
+
+    raw_example_dataset = ftrain_data.create_tf_dataset_for_client('1')
+    example_dataset = preprocess(raw_example_dataset, arg)
+    print(example_dataset.element_spec)
+
+    ftrain_data = make_federated_data(ftrain_data, np.arange(10), arg)
+
+    iterative_process = tff.learning.build_federated_averaging_process(
+        lambda: create_tff_model(
+            arg, len(in_vocab['vocab']), len(slot_vocab['vocab']),
+            len(intent_vocab['vocab']), example_dataset.element_spec),
+        client_optimizer_fn=lambda: tf.keras.optimizers.SGD(learning_rate=0.02
```

Comment:
Accept learning rates for both the client and the server as command line arguments.